server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -375,28 +391,57 @@ private WatchKey enableSettingsWatcher(WatchKey previousKey, Path settingsDir) t
         );
     }
 
-    CountDownLatch processFileSettings(Path path, Consumer<Exception> errorHandler) throws IOException {
+    CountDownLatch processFileSettings(Path path, Consumer<Exception> errorHandler) {
         CountDownLatch waitForCompletion = new CountDownLatch(1);
         logger.info("processing path [{}] for [{}]", path, NAMESPACE);
         try (
             var fis = Files.newInputStream(path);
             var bis = new BufferedInputStream(fis);
             var parser = JSON.xContent().createParser(XContentParserConfiguration.EMPTY, bis)
         ) {
-            stateService.process(NAMESPACE, parser, (e) -> {
-                try {
-                    if (e != null) {
-                        errorHandler.accept(e);
+            ReservedStateChunk parsedState = stateService.parse(NAMESPACE, parser);
+            if (nodeInfosRefreshRequired || nodesInfoResponse == null) {
+                var nodesInfoRequest = NodesInfoRequest.requestWithMetrics(NodesInfoRequest.Metric.INGEST);
+
+                clusterAdminClient().nodesInfo(nodesInfoRequest, new ActionListener<>() {
```

Comment:
Is `processFileSettings` protected by a lock or AtomicBoolean or anything like that? What happens if the file on disk is replaced hundreds of times every second, will we end up issuing many nodes info requests since we don't have any lock to ensure we only retrieve it once?

(I'm not super close to the internals of the file settings stuff, so it may be that this isn't an issue because we only invoke it once every X seconds, let me know if that's the case)

Reply:
This is a really good question, so we have only one thread that processes the file change events and the return of this call (`processFileSettings`) is a latch, the `CountDownLatch waitForCompletion`. The caller of process file settings waits on this latch, for all async actions to complete, before it decides to pick up another event and to check if the file changed since last time. This effectively makes the processing one at a time, until we have successfully written file state or hit an error we don't try to process again.

The relevant caller code is here: https://github.com/elastic/elasticsearch/pull/89735/files#diff-5a2c0417a5b4bdbcc6873fc853db7a8c531b47536bd41b84e20bb6c7b21ad6d9R334

We also don't trust the events we get from the JDK watch service that the file has changed, we only use them as an indicator that we should check if the file is different. If the file hasn't changed the OS file key or the modified timestamp we skip the event. The reason we can't rely on the JDK watcher is because the series of events that get sent are vastly different between OSs.